org.eclipse.paho.android.service/src/main/java/org/eclipse/paho/android/service/HandlerPingSender.java

```diff
@@ -0,0 +1,110 @@
+package org.eclipse.paho.android.service;
+
+import android.app.Service;
+import android.os.Handler;
+import android.os.PowerManager;
+import android.os.PowerManager.WakeLock;
+import android.util.Log;
+
+import org.eclipse.paho.client.mqttv3.IMqttActionListener;
+import org.eclipse.paho.client.mqttv3.IMqttToken;
+import org.eclipse.paho.client.mqttv3.MqttPingSender;
+import org.eclipse.paho.client.mqttv3.internal.ClientComms;
+
+import java.util.Date;
+
+/**
+ * Ping sender implementation using Android Handlers.
+ *
+ * <p>This class implements the {@link MqttPingSender} pinger interface
+ * allowing applications to send ping packet to server every keep alive interval.
+ * </p>
+ *
+ * @see MqttPingSender
+ */
+class HandlerPingSender implements MqttPingSender, Runnable {
+    // Identifier for Intents, log messages, etc..
+    private static final String TAG = "AlarmPingSender";
+    private final String wakeLockTag = MqttServiceConstants.PING_WAKELOCK + ":PingSender";
+    // TODO: Add log.
+    private ClientComms comms;
+    private MqttService service;
+    private Handler handler;
+    private WakeLock wakelock;
+    private volatile boolean hasStarted = false;
+    
+    public HandlerPingSender(MqttService service) {
+        if (service == null) {
+            throw new IllegalArgumentException(
+                "Neither service nor client can be null.");
+        }
+        this.service = service;
+        this.handler = new Handler();
+    }
+    
+    @Override
+    public void init(ClientComms comms) {
+        this.comms = comms;
+    }
+    
+    @Override
+    public void start() {
+        String action = MqttServiceConstants.PING_SENDER
+            + comms.getClient().getClientId();
+        Log.d(TAG, "Register handler to MqttService" + action);
+        
+        schedule(comms.getKeepAlive());
+        hasStarted = true;
+    }
+    
+    @Override
+    public void stop() {
+        Log.d(TAG, "Unregister handler to MqttService" + comms.getClient().getClientId());
+        if (hasStarted) {
+            handler.removeCallbacks(this);
+            
+            hasStarted = false;
+        }
+    }
+    
+    @Override
+    public void schedule(long delayInMilliseconds) {
+        Date nextSchedule = new Date(System.currentTimeMillis() + delayInMilliseconds);
+        Log.d(TAG, "Schedule next ping at " + nextSchedule.toString());
+        
+        handler.postDelayed(this, delayInMilliseconds);
```

Comment:
Huh, that's clever. Just start a new thread function on the interval. However within what thread context is this being started in, does it create a problem if it's started in the UI thread context?

Reply:
Hm. https://developer.android.com/training/multiple-threads/communicate-ui#Handler

https://developer.android.com/reference/android/os/Handler.html#Handler()
Looks like it'll make the handler on whatever thread constructs this object.  (Since that's where ```this.handler = new Handler();```) is called.

That might be an interesting bug. With this thing implementing runnable, I doubt the intention was for the handler to be run on the UI thread (if that's where this object is constructed) while the rest of the thread runs on a new thread (in the run() method)

Reply:
I took a peek up the call stack for where this class is constructed. Way up in MqttAndroidClient, we construct one of these on an `Executors.newCachedThreadPool()`. (from calling connect())

However, elsewhere in mqttAndroidClient, we end up calling this class' constructor within a ServiceConnection (run by context.bindService), so it might end up being called on the main thread in some cases. :man_shrugging: